utils/queue/priorityQueue.go

```diff
@@ -0,0 +1,84 @@
+package queue
+
+import (
+	"container/heap"
+	"time"
+)
+
+// Item stores the attributes which will be pushed to the priority queue..
+type Item[T any] struct {
+	Value     T
+	Priority  int
+	timeStamp int64
+	index     int
+}
+
+// PriorityQueue ....
+type PriorityQueue[T any] []*Item[T]
+
+// Len: Size of the priority queue . Used to satisfy the heap interface...
+func (pq PriorityQueue[T]) Len() int { return len(pq) }
+
+// Less is used to compare elements and store them in the proper order in
+// priority queue.
+func (pq PriorityQueue[T]) Less(i, j int) bool {
+	if pq[i].Priority == pq[j].Priority {
+		return pq[i].timeStamp <= pq[j].timeStamp
+	}
+	return pq[i].Priority > pq[j].Priority
+}
+
+// Swap is used to swap the values in the priority queue.
+func (pq PriorityQueue[T]) Swap(i, j int) {
+	pq[i], pq[j] = pq[j], pq[i]
+	pq[i].index = i
+	pq[j].index = j
+}
+
+// Push adds elements to the priority queue
+func (pq *PriorityQueue[T]) Push(x interface{}) {
+	n := len(*pq)
+	item := x.(*Item[T])
+	item.index = n
+	item.timeStamp = makeTimestamp()
+	*pq = append(*pq, item)
+}
+
+// Pop removes elements from the priority queue
+func (pq *PriorityQueue[T]) Pop() interface{} {
+	old := *pq
+	n := len(old)
+	if n == 0 {
+		return nil
+	}
+	item := old[n-1]
+	old[n-1] = nil
```

Comment:
Is this line necessary given that we'll reslice `pq`? :thinking:

Reply:
According to the [example](https://pkg.go.dev/container/heap#example-package-PriorityQueue)
```go
old[n-1] = nil  // avoid memory leak
item.index = -1 // for safety
```